File: .claude/hooks/pre-read-json-csv-compact.mjs
#!/usr/bin/env node
// Résume les grands fichiers JSON/CSV/JSONL avant lecture pour économiser le contexte (PreToolUse Read)
import { readFileSync, existsSync, statSync } from 'fs';
import { fileURLToPath } from 'url';
import { extname, basename } from 'path';

const BYTE_THRESHOLD = 50_000;
const CSV_PREVIEW_ROWS = 5;
const JSONL_PREVIEW_LINES = 3;

function inferJsonSchema(value, depth = 0) {
  if (depth > 2) return typeof value;
  if (value === null) return 'null';
  if (Array.isArray(value)) {
    const inner = value.length ? inferJsonSchema(value[0], depth + 1) : 'unknown';
    return `Array<${inner}> (${value.length} items)`;
  }
  if (typeof value === 'object') {
    const keys = Object.keys(value).slice(0, 12);
    const fields = keys.map((k) => `${k}: ${inferJsonSchema(value[k], depth + 1)}`);
    const extra = Object.keys(value).length > 12 ? ` … +${Object.keys(value).length - 12} more` : '';
    return `{ ${fields.join(', ')}${extra} }`;
  }
  return typeof value;
}

function summarizeJson(raw, filePath) {
  let parsed;
  try { parsed = JSON.parse(raw); } catch { return null; }

  const lines = raw.split('\n').length;
  const bytes = Buffer.byteLength(raw, 'utf8');
  const name = basename(filePath);

  if (Array.isArray(parsed)) {
    const schema = parsed.length ? inferJsonSchema(parsed[0]) : 'empty';
    return `[json-csv-compact] \`${name}\` — JSON array (${parsed.length} items, ${lines} lines, ${(bytes / 1024).toFixed(1)} KB)\nItem schema: ${schema}`;
  }

  const schema = inferJsonSchema(parsed);
  return `[json-csv-compact] \`${name}\` — JSON object (${lines} lines, ${(bytes / 1024).toFixed(1)} KB)\nSchema: ${schema}`;
}

function summarizeCsv(raw, filePath) {
  const lines = raw.split('\n').filter((l) => l.trim());
  if (!lines.length) return null;

  const name = basename(filePath);
  const headers = lines[0];
  const preview = lines.slice(1, 1 + CSV_PREVIEW_ROWS);
  const total = lines.length - 1;

  const cols = headers.split(',').length;
  const rows = [headers, ...preview].join('\n');
  return `[json-csv-compact] \`${name}\` — CSV (${total} rows, ${cols} columns)\nFirst ${preview.length} rows:\n\`\`\`\n${rows}\n\`\`\``;
}

function summarizeJsonl(raw, filePath) {
  const lines = raw.split('\n').filter((l) => l.trim());
  if (!lines.length) return null;

  const name = basename(filePath);
  const preview = lines.slice(0, JSONL_PREVIEW_LINES).map((l) => {
    try { return inferJsonSchema(JSON.parse(l)); } catch { return l.slice(0, 80); }
  });

  return `[json-csv-compact] \`${name}\` — JSONL (${lines.length} lines)\nFirst ${preview.length} line schemas:\n${preview.map((s, i) => `  ${i + 1}. ${s}`).join('\n')}`;
}

export function run(input, { readFile = readFileSync, exists = existsSync, stat = statSync } = {}) {
  if (input.tool_name !== 'Read') return null;

  const filePath = input.tool_input?.file_path ?? '';
  if (!filePath) return null;

  const ext = extname(filePath).toLowerCase().replace('.', '');
  if (!['json', 'csv', 'jsonl', 'tsv', 'ndjson'].includes(ext)) return null;

  if (!exists(filePath)) return null;

  let size;
  try { size = stat(filePath).size; } catch { return null; }
  if (size < BYTE_THRESHOLD) return null;

  let raw;
  try { raw = readFile(filePath, 'utf8'); } catch { return null; }

  let summary;
  if (ext === 'json') summary = summarizeJson(raw, filePath);
  else if (ext === 'csv' || ext === 'tsv') summary = summarizeCsv(raw, filePath);
  else summary = summarizeJsonl(raw, filePath);

  if (!summary) return null;

  return { decision: 'block', reason: summary };
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/pre-read-json-csv-compact.py
#!/usr/bin/env python3
# @hookstack pre-read-json-csv-compact
"""Summarizes large JSON/CSV/JSONL files before reading (PreToolUse Read)."""
import json
import os
import sys

BYTE_THRESHOLD = 50_000
CSV_PREVIEW_ROWS = 5
JSONL_PREVIEW_LINES = 3
EXTENSIONS = {"json", "csv", "jsonl", "tsv", "ndjson"}


def _read_file(path):
    with open(path, "r", encoding="utf8") as f:
        return f.read()


def _exists(path):
    return os.path.exists(path)


def _stat_size(path):
    return os.stat(path).st_size


def infer_json_schema(value, depth=0):
    if depth > 2:
        return type(value).__name__
    if value is None:
        return "null"
    if isinstance(value, list):
        inner = infer_json_schema(value[0], depth + 1) if value else "unknown"
        return f"Array<{inner}> ({len(value)} items)"
    if isinstance(value, dict):
        keys = list(value.keys())[:12]
        fields = ", ".join(f"{k}: {infer_json_schema(value[k], depth + 1)}" for k in keys)
        extra = f" … +{len(value) - 12} more" if len(value) > 12 else ""
        return f"{{ {fields}{extra} }}"
    return type(value).__name__


def summarize_json(raw, file_path):
    try:
        parsed = json.loads(raw)
    except Exception:
        return None

    lines = len(raw.split("\n"))
    bytes_n = len(raw.encode("utf-8"))
    name = os.path.basename(file_path)

    if isinstance(parsed, list):
        schema = infer_json_schema(parsed[0]) if parsed else "empty"
        return (
            f"[json-csv-compact] `{name}` — JSON array ({len(parsed)} items, "
            f"{lines} lines, {bytes_n / 1024:.1f} KB)\nItem schema: {schema}"
        )

    schema = infer_json_schema(parsed)
    return (
        f"[json-csv-compact] `{name}` — JSON object ({lines} lines, "
        f"{bytes_n / 1024:.1f} KB)\nSchema: {schema}"
    )


def summarize_csv(raw, file_path):
    lines = [l for l in raw.split("\n") if l.strip()]
    if not lines:
        return None

    name = os.path.basename(file_path)
    headers = lines[0]
    preview = lines[1 : 1 + CSV_PREVIEW_ROWS]
    total = len(lines) - 1

    cols = len(headers.split(","))
    rows = "\n".join([headers] + preview)
    return (
        f"[json-csv-compact] `{name}` — CSV ({total} rows, {cols} columns)\n"
        f"First {len(preview)} rows:\n```\n{rows}\n```"
    )


def summarize_jsonl(raw, file_path):
    lines = [l for l in raw.split("\n") if l.strip()]
    if not lines:
        return None

    name = os.path.basename(file_path)
    preview = []
    for l in lines[:JSONL_PREVIEW_LINES]:
        try:
            preview.append(infer_json_schema(json.loads(l)))
        except Exception:
            preview.append(l[:80])

    schemas = "\n".join(f"  {i}. {s}" for i, s in enumerate(preview, start=1))
    return (
        f"[json-csv-compact] `{name}` — JSONL ({len(lines)} lines)\n"
        f"First {len(preview)} line schemas:\n{schemas}"
    )


def run(input_data, *, read_file=None, exists=None, stat_size=None):
    if read_file is None:
        read_file = _read_file
    if exists is None:
        exists = _exists
    if stat_size is None:
        stat_size = _stat_size

    if input_data.get("tool_name") != "Read":
        return None

    file_path = (input_data.get("tool_input") or {}).get("file_path") or ""
    if not file_path:
        return None

    ext = os.path.splitext(file_path)[1].lower().lstrip(".")
    if ext not in EXTENSIONS:
        return None

    if not exists(file_path):
        return None

    try:
        size = stat_size(file_path)
    except Exception:
        return None
    if size < BYTE_THRESHOLD:
        return None

    try:
        raw = read_file(file_path)
    except Exception:
        return None

    if ext == "json":
        summary = summarize_json(raw, file_path)
    elif ext in ("csv", "tsv"):
        summary = summarize_csv(raw, file_path)
    else:
        summary = summarize_jsonl(raw, file_path)

    if not summary:
        return None

    return {"decision": "block", "reason": summary}


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
